docs/writing_rules/examples.md

```diff
@@ -0,0 +1,232 @@
+subprocess.run("grep -R {} .".format(sys.argv[1]), shell=True) # Doesn't match here
+```
+
+https://semgrep.live/d8J6
+
+Semgrep will match `(..., shell=True)` only when `shell=True` is the last argument. To fix this, we can use the ellipsis operator on both sides of `shell=True`.
+
+```yaml
+patterns:
+- pattern-not: subprocess.call("...")
+- pattern: subprocess.call(..., shell=True, ...)
+```
+
+```python
+import subprocess
+import sys
+
+subprocess.call("echo 'hello'") # Doesn't match
+subprocess.call("grep -R {} .".format(sys.argv[1])) # Doesn't match
+subprocess.call("grep -R {} .".format(sys.argv[1]), shell=True) # Matches here
+subprocess.call("grep -R {} .".format(sys.argv[1]), shell=True, cwd="/home/user") # Matches here too!
+subprocess.run("grep -R {} .".format(sys.argv[1]), shell=True) # Doesn't match here
+```
+
+https://semgrep.live/ZqKW
+
+**Bonus: Match any `subprocess` function with `shell=True`.** As you probably noticed, `subprocess.run` is subject to the same issue as `subprocess.call`. `subprocess.run` [was made available in Python 3.5](https://docs.python.org/3/library/subprocess.html#older-high-level-api). We can match both `subprocess.call` and `subprocess.run` by using **metavariables**. Metavariables let you match any code expression. To use metavariables in Semgrep, use the dollar sign as a prefix and all capital letters. In this example, we will use `subprocess.$FUNC`. The name can be anything -- it's just a like a variable in a normal language and will "hold" the expression it matches.
+
+```yaml
+patterns:
+- pattern-not: subprocess.$FUNC("...")
+- pattern: subprocess.$FUNC(..., shell=True, ...)
+```
+
+```python
+import subprocess
+import sys
+
+subprocess.call("echo 'hello'") # Doesn't match
+subprocess.call("grep -R {} .".format(sys.argv[1])) # Doesn't match
+subprocess.call("grep -R {} .".format(sys.argv[1]), shell=True) # Matches here
+subprocess.call("grep -R {} .".format(sys.argv[1]), shell=True, cwd="/home/user") # Matches here
+subprocess.run("grep -R {} .".format(sys.argv[1]), shell=True) # Matches here too!
+```
+
+https://semgrep.live/nJ9d
+
+## Enforce Specific Use of an API
+
+Sometimes you may wish to enforce the specific use of a function's API. There are many examples of this, such as `subprocess.call(..., shell=True, ...)` above; you may wish to match and fail any commit where `shell=True`. This is easy to do in Semgrep, as seen in the above section.
+
+However, there are also function APIs that are insecure by default--or insecure depending on context, such as Jinja2, which does [not enable autoescaping by default](https://github.com/pallets/jinja/blob/2a8515d2e53a2be475d9df3fe44e308501201a95/src/jinja2/environment.py#L296). Jinja2 is an arbitrary templating engine, so this makes sense in non-web contexts. This may not be obvious, though, and if you are working directly with the Jinja2 engine in a web context you want to make sure `autoescape=True`.
```

Comment:
Propose removing "function APIs" => "APIs"

Add "," after "web context"

"to make sure" => should this be "to make sure to include"?